File: leetcode-215/answer-quick sort.py
from typing import List
from random import randrange


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        self.quickSelect(nums, 0, len(nums)-1, k)
        return nums[k-1]

    def partition(self, _nums: List[int], _left: int, _right: int) -> int:
        randomInt = randrange(_left, _right+1)
        pivot = _nums[randomInt]
        _nums[randomInt] = _nums[_left]
        _nums[_left] = pivot
        while(_left < _right):
            while(_left < _right and _nums[_right] <= pivot):
                _right -= 1
            _nums[_left] = _nums[_right]
            while(_left < _right and _nums[_left] >= pivot):
                _left += 1
            _nums[_right] = _nums[_left]
        _nums[_left] = pivot
        return _left

    def quickSelect(self, _nums: List[int], _left: int, _right: int, index: int) -> None:
        n = self.partition(_nums, _left, _right)
        if n < index-1:
            self.quickSelect(_nums, n+1, _right, index)
        elif n > index-1:
            self.quickSelect(_nums, _left, n-1, index)

File: leetcode-215/answer-quick-sort.cpp
#include <vector>
#include <random>

using namespace std;

class Solution
{
    int partition(vector<int> &_nums, int _left, int _right)
    {
        static default_random_engine e;
        uniform_int_distribution<int> u(_left, _right); //should not be static here
        int rand = u(e);
        int pivot = _nums[rand];
        _nums[rand] = _nums[_left];
        _nums[_left] = pivot;
        while (_left < _right)
        {
            while (_left < _right && _nums[_right] <= pivot)
                --_right;
            _nums[_left] = _nums[_right];
            while (_left < _right && _nums[_left] >= pivot)
                ++_left;
            _nums[_right] = _nums[_left];
        }
        _nums[_left] = pivot;
        return _left;
    }

    void quickSelect(vector<int> &_nums, int _left, int _right, int index)
    {
        int n = partition(_nums, _left, _right);
        if (n > index - 1)
            quickSelect(_nums, _left, n - 1, index);
        else if (n < index - 1)
            quickSelect(_nums, n + 1, _right, index);
    }

public:
    int findKthLargest(vector<int> &nums, int k)
    {
        quickSelect(nums, 0, nums.size() - 1, k);
        return nums[k - 1];
    }
};
